src/Database/LSMTree/Internal/MergingTree.hs

```diff
@@ -272,3 +295,42 @@ finalise mergeState = releaseMTS =<< readMVar mergeState
     releasePER (PreExistingRun         r) = releaseRef r
     releasePER (PreExistingMergingRun mr) = releaseRef mr
 
+{-# SPECIALISE remainingMergeDebt ::
+     Ref (MergingTree IO h) -> IO (MergeDebt, NumEntries) #-}
+-- | Calculate an upper bound on the merge credits required to complete the
+-- merge, i.e. turn the tree into a 'CompletedTreeMerge'. For the recursive
+-- calculation, we also return an upper bound on the size of the resulting run.
+remainingMergeDebt ::
+     (MonadMVar m, PrimMonad m)
+  => Ref (MergingTree m h) -> m (MergeDebt, NumEntries)
+remainingMergeDebt (DeRef mt) = do
+    readMVar (mergeState mt) >>= \case
+      CompletedTreeMerge r -> return (MergeDebt 0, Run.size r)
+      OngoingTreeMerge mr  -> MR.remainingMergeDebt mr
+      PendingTreeMerge ptm -> remainingMergeDebtPendingMerge ptm
+
+{-# SPECIALISE remainingMergeDebtPendingMerge ::
+     PendingMerge IO h -> IO (MergeDebt, NumEntries) #-}
+remainingMergeDebtPendingMerge ::
+     (MonadMVar m, PrimMonad m)
+  => PendingMerge m h -> m (MergeDebt, NumEntries)
+remainingMergeDebtPendingMerge (PendingMerge _ prs mts) = do
+    -- TODO: optimise to reduce allocations
+    debtsPre <- traverse remainingMergeDebtPreExistingRun prs
+    debtsChild <- traverse remainingMergeDebt mts
+    return (mergeDebt (debtsPre <> debtsChild))
+  where
+    remainingMergeDebtPreExistingRun = \case
+        PreExistingRun        r  -> return (MergeDebt 0, Run.size r)
+        PreExistingMergingRun mr -> MR.remainingMergeDebt mr
+
+mergeDebt :: Foldable f => f (MergeDebt, NumEntries) -> (MergeDebt, NumEntries)
+mergeDebt debts =
```

Comment:
might be better to call this `sumMergeDebt` or something. Otherwise it can be confused with being the same concept as `MergeDebt`.

Reply:
`sum` doesn't quite feel right, because it's not just adding up the debts and sizes of the child merges, it also includes the debt of the new merge. So I called it `debtOfNestedMerge` now.